### ISSUE TITLE
VS 2015 Update 2 package update hangs while resolving dependencies

### ISSUE THREAD

**Issue by chrisdrobison** [NuGet/Home#2420]
I just updated to VS 2015 Update 2 and the latest Nuget 3.4.1. I am trying to upgrade a package for just one project and it hangs after the following message:

```
Attempting to gather dependency information for package 'Common.Logging.3.3.1' with respect to project 'Utilities\PriceListImport', targeting '.NETFramework,Version=v4.5'
Attempting to resolve dependencies for package 'Common.Logging.3.3.1' with DependencyBehavior 'Lowest'
```

The following is our package.json

```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Common.Logging" version="3.1.0" targetFramework="net45" />
  <package id="Common.Logging.Core" version="3.1.0" targetFramework="net45" />
  <package id="Common.Logging.NLog32" version="3.1.0" targetFramework="net45" />
  <package id="Core.App.Database" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Busi" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Busi.SmartObjects" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Busi.XactNet" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Busi.XCCollab" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Data" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Data.ContentsScope" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Data.Enums" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Data.Imports" version="1.0" targetFramework="net45" />
  <package id="Core.Data.Res" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Data.ThirdPartyApps" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Data.XCCollab" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Db.Database" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Db.Tables" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Interface.IDatabase" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Interface.ITables" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.MP" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.MP.XactNet" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Rpt.Data" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Rpt.Print" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Rpt.Res" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Sys" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Sys.Drawing" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Trail" version="1.0.0.0" targetFramework="net45" />
  <package id="Core.UI" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.UI.Converters" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.UI.Extended" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.UI.Reporting" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.UI.WPF" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.UI.XactNet" version="28.0.0.31115" targetFramework="net45" />
  <package id="Core.Win32" version="28.0.0.31115" targetFramework="net45" />
  <package id="CsvHelper" version="2.13.2.0" targetFramework="net45" />
  <package id="Elasticsearch.Net" version="1.7.0" targetFramework="net45" />
  <package id="Extended.Wpf.Toolkit" version="2.0.0" targetFramework="net45" />
  <package id="NEST" version="1.7.0" targetFramework="net45" />
  <package id="Newtonsoft.Json" version="7.0.1" targetFramework="net45" />
  <package id="NLog" version="3.2.1" targetFramework="net45" />
  <package id="NLog.Config" version="3.2.1" targetFramework="net45" />
  <package id="NLog.Schema" version="3.2.1" targetFramework="net45" />
  <package id="Ookii.Dialogs" version="1.0" targetFramework="net45" />
  <package id="Sketch.Data.Res" version="28.0.0.31115" targetFramework="net45" />
  <package id="Sketch.Interface" version="28.0.0.31115" targetFramework="net45" />
  <package id="Sketch.MP.Settings" version="28.0.0.31115" targetFramework="net45" />
  <package id="SSH.NET" version="2013.4.7" targetFramework="net45" />
  <package id="Xceed.Editors" version="2.1.4337.22935" targetFramework="net45" />
  <package id="Xceed.Grid" version="3.2.4337.22936" targetFramework="net45" />
  <package id="Xceed.UI" version="1.1.4337.22935" targetFramework="net45" />
  <package id="XceedZip" version="2.0.110.0" targetFramework="net45" />
  <package id="Xm8.App.FieldApp" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.DataTransfer" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.Dim" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.EDPL" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.EstimateData" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.Monetary" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.PaymentTracker" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.PL" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.PLData" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.ProjAccum" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.ProjComparison" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.ProjMan" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.ProjMerge" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.ProjReprice" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.Search" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.ServerSide" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.Transfer" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Busi.WorkOrder" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.ClaimDoc" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.Contact" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.InspRes" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.PLDoc" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.ProjCalcFormula" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.ProjDoc" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.ProjPLShared" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.Res" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.SystemDoc" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.WorkOrderDoc" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.XV" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Data.XVRes" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Db.Tables" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Interface.ITables" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.ClaimManager" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.Contact" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.CovLoss" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.DataTransfer" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.EWD" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.Global" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.GlobalChanges" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.Low" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.Proj" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.ProjCalcFormula" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.ProjMan" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.ProjPLShared" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.QualityReviewWizard" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.MP.XactNet" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Rpt.ContactAB" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Rpt.DataParse" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Rpt.DataReports" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Rpt.PLAB" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Rpt.ProjAB" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.Rpt.XVAB" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.UI.Contact" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.UI.DataTransfer" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.UI.Global" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.UI.Low" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.UI.PrintLib" version="28.0.0.31115" targetFramework="net45" />
  <package id="Xm8.UI.ProjMergeWizard" version="28.0.0.31115" targetFramework="net45" />
</packages>
```

Most of those packages are from an internal Nuget repo. Not sure what to do. It works for some projects, but not others.

**Comment by rrelyea** [comment]
Resolving shouldn't have changed from 3.3
How long did you wait before you gave up on the "hang"?
What happens if you wait longer?

**Comment by yishaigalatzer** [comment]
Can you tell us a bit more about your project?
What kind of internal nuget server are you running?
How many versions are available for each of these packages on the server?
Are all these packages depending on common.logging?

Did this behave differently on nuget 3.3? Can you rollback and compare?

**Comment by chrisdrobison** [comment]
What is the best way to rollback on update 2?

**Comment by rrelyea** [comment]
try this:
Tools/Extensions and Updates>Installed>All>Nuget Package Manager for Visual Studio 2015>Uninstall
Then install 3.3 for vs2015 from: http://dist.nuget.org/index.html

**Comment by Myfreedom614** [comment]
@rrelyea for 3.4.1 version, after upgrading to VS 2015 update 2, the Available package sources will be empty and can't add custom source in the options, this works but this should not be a workaround. For 3.4.1 version, the work around is to delete the C:\Users{username}\AppData\Roaming\NuGet\nuget.config

**Comment by chrisdrobison** [comment]
I'm trying this now, but to answer your other questions, we use Sonatype's Nexus as our internal repo. For the internal packages, there are 8 versions available for each. None of them, depend on Common.Logging.

**Comment by chrisdrobison** [comment]
I've down graded and it seems to be exhibiting the same behavior. I'm at 30 minutes and counting. VS is spinning at 25% CPU.

**Comment by chrisdrobison** [comment]
It's been spinning on the same project for an hour now. I need to get some work done so I'll restart this process before I leave and see if anything comes back during the night.

**Comment by yishaigalatzer** [comment]
@chrisdrobison thanks for getting back with all the details. The good news is we know that this is not a regression (phew...).

The other good news, is that this is now our highest priority item, is there a chance to put our hands on your project so we can work on reproducing, getting the core reason why it is so slow, and testing some new algorithms we have in mind?

**Comment by chrisdrobison** [comment]
I think we can work something out. Reach out to me at my work email: cdrobison@xactware.com.

**Comment by yishaigalatzer** [comment]
Cool! Assigned a developer, he should reach out to you shortly

**Comment by xperiandri** [comment]
I have an issue similar to @Myfreedom614 
In 3.4.1.807 search does not work in UI.
When I scroll on browse tab it loads more packages. But as soon as I put anything into search box it either finds a few packages (for example if to search for system) or nothing (search for Prism or System.Collections.Immutable)

**Comment by rrelyea** [comment]
@Myfreedom614  @xperiandri  - neither of your issues are the same as this issue, I believe. Can you please open a new issue detailing what you are seeing?

**Comment by xperiandri** [comment]
#2429

**Comment by yishaigalatzer** [comment]
@chrisdrobison thanks for sharing your packages, we found a bug in the update code that will make it spin forever when the package graph is super deep (like yours, it is over 20 packages deep). We are working on a fix. 

Much much appreciated!

**Comment by chrisdrobison** [comment]
@yishaigalatzer Glad I could help.

**Comment by yishaigalatzer** [comment]
Hey @chrisdrobison we now have a fix ready. Would you be able to send us your private email address so we can put private bits in your hands for early validation (Same stands for anyone else interested in trying out the fix).

Please use the following two email addresses:

yigalatz @ microsoft.com
juste @ microsoft.com

**Comment by jainaashish** [comment]
Commit with PR https://github.com/NuGet/NuGet.Client/pull/457 and also merged to release-3.4.2 and release-3.4.2-dev15
